transformer_engine/jax/csrc/extensions.cpp

```diff
@@ -43,6 +44,11 @@ pybind11::dict Registrations() {
         EncapsulateFunction(ScaledUpperTriangMaskedSoftmaxForward);
     dict["te_scaled_upper_triang_masked_softmax_backward"] =
         EncapsulateFunction(ScaledUpperTriangMaskedSoftmaxBackward);
+    dict["te_self_fused_attn_max_512_forward"] = EncapsulateFunction(SelfFusedAttnMax512Forward);
+    dict["te_self_fused_attn_max_512_backward"] = EncapsulateFunction(SelfFusedAttnMax512Backward);
+    dict["te_cross_fused_attn_max_512_forward"] = EncapsulateFunction(CrossFusedAttnMax512Forward);
+    dict["te_cross_fused_attn_max_512_backward"] =
+        EncapsulateFunction(CrossFusedAttnMax512Backward);
```

Comment:
Isn't `max_512` an internal implementation detail? I suggest we expose general functions in the API, and internally it will dispatch to the `max_512` functions:
```suggestion
    dict["te_self_fused_attn_forward"] = EncapsulateFunction(SelfFusedAttnForward);
    dict["te_self_fused_attn_backward"] = EncapsulateFunction(SelfFusedAttnBackward);
    dict["te_cross_fused_attn_forward"] = EncapsulateFunction(CrossFusedAttnForward);
    dict["te_cross_fused_attn_backward"] =
        EncapsulateFunction(CrossFusedAttnBackward);
```
This will make things easier once longer sequence length versions are ready. This dispatch function can either be in `common` or here in `jax/csrc`, whichever is more convenient. I wonder if the FP8 versions could also be incorporated, although handling all the FP8 stuff might make it messier than it's worth.

Reply:
The custom_calls ([cpp_extenions.py](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/jax/cpp_extensions.py), [modules.cpp](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/jax/csrc/modules.cpp)) are not intended to public to the users. From JAX side, we only expose the [module level APIs](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/api/jax.html) to public, that is people will only use `TransformerLayer` or `MultiheadAttention` module and the different fused attention implementation will be encapsulated in https://github.com/NVIDIA/TransformerEngine/blob/d717f535ea0c9c50ef426c6419182791ea9e9092/transformer_engine/jax/fused_attn.py#L46
(self_fused_attn is also not public and it is only used internally by the TE JAX)

The current design is to have [different custom calls](https://github.com/NVIDIA/TransformerEngine/blob/d717f535ea0c9c50ef426c6419182791ea9e9092/transformer_engine/jax/cpp_extensions.py#L2393-L2396) for different fused attention implementations and dispatching them on the python side ([fused_attn.py](https://github.com/NVIDIA/TransformerEngine/blob/d717f535ea0c9c50ef426c6419182791ea9e9092/transformer_engine/jax/fused_attn.py)). The sugestion that setting the dispatch functions on the C side is good, but I'm afraid it is not supported by JAX custom calls since the I/O buffers are static in JAX custom calls. The variable length of `aux_output_tensors` is not compatiable with the JAX custom calls.